### PR TITLE
fix(protocol-designer): grab defUri from labwareDefinitions key

### PR DESCRIPTION
**Pull request by jerader**
addresses RESC-161 https://opentrons.atlassian.net/browse/RESC-161

# Overview

There was a bug reported from the Support team where a Protocol made from version 6.0.0 was imported into 7.0.0 and then reimported into 7.0.0 and white screened. This is because there was a mistake in `loadLabware` in the 7.0.0 migration where the `loadName` was accidentally set to the `definitionURI` And then resulted in very brittle splitting logic in the `labwareInvariantProperties` reducer to grab the definition URI either from the <v7 `loadName` or from splitting the =v7 `labwareId`.

# Test Plan

Upload the `Created 8 September` protocol attached to the ticket. See that it imports correctly. Export it and then change the software version to "7.1.0" and reimport that exported protocol. It should work as expected.

# Changelog

- add to te 7.1.0 migration to change the loadName param in `loadLabware` commands to the true `loadName` instead of the `definitionId` (which is actually the definition uri but in schema 6 the top level labware key refers to it was "definitionId".).
- change the logic in the reducer to not branch and convert that loadname into the definition uri. 

# Review requests

see test plan

# Risk assessment

low

### CODE REVIEW

File: protocol-designer/src/load-file/migration/7_1_0.ts [hunk 1]
```diff
@@ -122,7 +122,7 @@ export const migrateFile = (
       },
     },
     labwareDefinitions: {
-      ...{ [trashId]: trashDefinition },
+      ...{ [trashDefUri]: trashDefinition },
```
Comment (jerader) [inline protocol-designer/src/load-file/migration/7_1_0.ts:160]: accidentally used `id` instead of `defUri` for the key here

File: protocol-designer/src/step-forms/reducers/index.ts [hunk 2]
```diff
@@ -1154,12 +1154,23 @@ export const labwareInvariantProperties: Reducer<
...
+            const labwareDefintionForV7 = Object.entries(
+              file.labwareDefinitions
+            ).find(([key, value]) => value.parameters.loadName === loadName)
+            let defUri: string = ''
+            if (labwareDefintionForV7 != null) {
+              const [key] = labwareDefintionForV7
+              defUri = key
+            } else {
+              defUri = loadName
+            }
```
Comment (jerader) [inline protocol-designer/src/step-forms/reducers/index.ts:1166]: if you import a protocol from v6 or lower, the `loadName` === `defUri`. But if you import a protocol from v7, the `loadName` === `labwareDefinition`'s `loadName`. That is why this logic is needed.

File: protocol-designer/src/load-file/migration/7_1_0.ts [hunk 3]
```diff
@@ -68,23 +68,29 @@ export const migrateFile = (
...
         return {
           ...stepForm,
-          blowout_location:
-            stepForm.blowout_location === 'fixedTrash'
-              ? trashId
-              : stepForm.blowout_location,
-          dropTip_location: trashId,
+          labware:
+            stepForm.labware === 'fixedTrash' ? trashId : stepForm.labware,
+          ...sharedParams,
```
Comment (jerader) [inline protocol-designer/src/load-file/migration/7_1_0.ts:93]: This is not related to the fix. This change is because when I imported the "created september 2022" protocol, lots of the transfer commands weren't being imported properly. It is because I forgot to migrate the `dispense_labware` and `labware` params in `moveLiquid` and `mix` from `fixedTrash` to the newly created trash entity's id. LMK if I should pull this out into its own PR though!

File: protocol-designer/src/load-file/migration/7_1_0.ts [hunk 4]
```diff
@@ -101,6 +107,22 @@ export const migrateFile = (
     filteredSavedStepForms
   )
 
+  const loadLabwareCommands: LoadLabwareCreateCommand[] = commands
+    .filter(
+      (command): command is LoadLabwareCreateCommand =>
+        command.commandType === 'loadLabware'
+    )
+    .map(command => {
+      const { parameters } = labwareDefinitions[command.params.loadName]
```
Comment (jerader) [inline protocol-designer/src/load-file/migration/7_1_0.ts:116]: this works because the loadName in the 7.0.0 migration is set to the definition URI.

File: protocol-designer/src/load-file/migration/7_1_0.ts [hunk 5]
```diff
@@ -68,23 +68,33 @@ export const migrateFile = (
...
         return {
           ...stepForm,
-          blowout_location:
-            stepForm.blowout_location === 'fixedTrash'
-              ? trashId
-              : stepForm.blowout_location,
-          dropTip_location: trashId,
+          labware:
+            stepForm.labware === 'fixedTrash' ? trashId : stepForm.labware,
+          ...sharedParams,
```
Comment (jerader) [inline protocol-designer/src/load-file/migration/7_1_0.ts:97]: This is not related to the fix. This change is because when I imported the "created september 2022" protocol, lots of the transfer commands weren't being imported properly. It is because I forgot to migrate the `dispense_labware`, `aspirate_labware`, and `labware` params in `moveLiquid` and `mix` from `fixedTrash` to the newly created trash entity's id (As part of the movable trash work). LMK if I should pull this out into its own PR though!

File: protocol-designer/src/step-forms/reducers/index.ts [hunk 6]
```diff
@@ -1154,12 +1154,19 @@ export const labwareInvariantProperties: Reducer<
         ...loadLabwareCommands.reduce(
           (acc: NormalizedLabwareById, command: LoadLabwareCreateCommand) => {
             const { labwareId, loadName } = command.params
-            const defUri = labwareId?.split(':')[1]
+            const labwareDefinitionMatch = Object.entries(
+              file.labwareDefinitions
+            ).find(
+              ([definitionUri, labwareDef]) =>
+                labwareDef.parameters.loadName === loadName
```
Comment (b-cooper) [inline protocol-designer/src/step-forms/reducers/index.ts]: can we match on the def uri here instead? That way we account for version and namespace differences

Comment (jerader) [reply protocol-designer/src/step-forms/reducers/index.ts]: how would we get the def uri though? I thought there is no other area in a file that has the defUri by itself except for the `labwareDefinitions` keys

Comment (b-cooper) [inline protocol-designer/src/step-forms/reducers/index.ts]: can we match on the def uri here instead? That way we account for version and namespace differences

Comment (b-cooper) [reply protocol-designer/src/step-forms/reducers/index.ts]: You can generate the def URI using the `getLabwareDefURI` utility function. it takes the definition in and concatenates the loadname, namespace and version

**Review by b-cooper — APPROVED** [review]
if we can match on defURI, this looks great

File: protocol-designer/src/step-forms/reducers/index.ts (hunk 6 above)
Comment (jerader) [reply protocol-designer/src/step-forms/reducers/index.ts]: cool thanks, ya i can do that!
